syntax = "proto2";

package dafs.proto;

message Delta
{
    required string difference = 1;
}

message Identity
{
    required string id = 1;
}

message BlockFormat
{
    required string contents = 1;
}

message BlockInfo
{
    required string path = 1;
    required Identity identity = 2;
    required int32 revision = 3;
}

message Endpoint
{
    required Address management = 1;
    required Address replication = 2;
    required Identity identity = 3;
}

message MetaData
{
    enum MetaDataType
    {
        BlockInfoKey = 0;
        BlockFormatKey = 1;
        AddressKey = 2;
        EndpointKey = 3;
        IdentifierKey = 4;
        NodeEndpointsKey = 5;
        MinusReplicatedEndpointsKey = 6;
        ZeroReplicatedEndpointsKey = 7;
        PlusReplicatedEndpointsKey = 8;
        MinusIdentityKey = 9;
        ZeroIdentityKey = 10;
        PlusIdentityKey = 11;
    }
    required MetaDataType type = 1;
    required string value = 2;
}

message Address
{
    required string ip = 1;
    required int32 port = 2;
}

message Message
{
    enum MessageType
    {
        SUCCESS = 0;
        FAILURE = 1;
        ALLOCATE_BLOCK = 2;
        READ_BLOCK = 3;
        WRITE_BLOCK = 4;
        DELETE_BLOCK = 5;
        GET_NODE_DETAILS = 6;
        JOIN_CLUSTER = 7;
        REQUEST_JOIN_CLUSTER = 8;
        ACCEPT_JOIN_CLUSTER = 9;
        EXIT_CLUSTER = 10;
        PROPOSE_EXIT_CLUSTER = 11;
        MINUS_EXIT_CLUSTER = 12;
    }
    required Address from_address = 1;
    required Address to_address = 2;
    required MessageType type = 3;
    repeated MetaData metadata = 4;
}
